### ISSUE TITLE
Adding new arguments while editing a JSON-schema-based WorkflowTask

### ISSUE THREAD

**Issue by tcompa** [fractal-analytics-platform/fractal-web#128]
Branching from #82 

> > The first step towards this feature should concern with scalar arguments (that is, no objects and no arrays). This is because objects and arrays are slightly more complex to handle (we should be able to add/remove elements, since their number is not defined).     
> 
> When we tackle 7 (adding lists, object content), it automatically means allowing addition of some arguments in the schema example (though limited), so that seems to be the point for generalization of these 2 approaches to me

I think we have a misunderstanding, as I see no automatic connection between these two use cases:
1. Adding an item to an "array" argument is perfectly in line with the schema definition. The schema says that argument `X` is a array of integers, and does not include one entry for each element. From the web client, we can add as many items as we'd like, and we never have to update the schema.
2. Adding a new argument (e.g. argument `Y`, which doesn't exist in the pydantic-generated schema) means changing the schema. I guess we should not modify the schema in the DB (the one being stored at, say, `Task.args_schema`), because that one is shared by all `WorkflowTask`s based on a given `Task`; what we may do is to make a local copy of the schema inside the web client (associated to the current WorkflowTask), which we can modify. But if this information is not fed back to the DB, there may still be a missing step.

To reiterate what I said: use case 2 also implies that the task cannot use pydantic to validate arguments: If it does, then the validation will strip argument `Y` from the arguments, since it is not known.
A possible idea to move forward (*) is the following:
* Move away from pydantic for validation, and only rely on the (pydantic-generated) schemas.
* If we go in this direction, validation could also be done in fractal-server (rather than doing it in fractal-tasks-core), whenever a task has an associated schema (in the DB).
* At that point, we could also decide that schemas can be modified (either at the `Task` or `WorkflowTask` level), and this will have the desired effects: if I add argument `Y` to the schema, this information will enter the DB, and validation (that will be performed in fractal-server) will correctly accept argument `Y`.

(*) I'm not proposing we follow this route, and if we do then we need to discuss it in much more detail. This is mostly an attempt to make things more explicit.

**Comment by jluethi** [comment]
Hmm, let's make sure we don't make this bigger than it needs to be. I fully agree that adding new arguments to a task with a json schema should not be a typical use-case. The only user story I can think of for this is task development and making it easy if a user modifies the task python file to run.

For example:
As a power-user, I want to play around making a custom version of the cellpose task that also takes arguments A & B, on top of all the others. Intuitively, I'd just modify the cellpose task python file => the actual task & the pydantic validation.

=> I updated the pydantic validation in the class and then I'm confused why they don't show up in the web interface (=> because I did not update the json schema in the database). Given the modified Python source code, the actual task can take more arguments and it would pass the pydantic check.

----

The scenario above clearly is an edge case, but one that can arise during task development. I see 2 ways of approaching this:
1. We don't allow entries to be added to the main argument list. The user needs to update the schema for those to show up. Slightly trickier to go through the scenario above, but not a problem often. Also, may be confusing to users why they can add e.g. extra channels, but not extra top-level arguments
2. We allow the user to add arguments. But if they add something that's not in the Pydantic check, their task will fail upon running it.

I can see why 2 may be the more annoying scenario, e.g. something that even non power user can trigger accidentally.


> To reiterate what I said: use case 2 also implies that the task cannot use pydantic to validate arguments: If it does, then the validation will strip argument Y from the arguments, since it is not known.

I like the pydantic validation as part of the tasks, because then our command-line executables themselves have input validation.

----

tldr: Adding new arguments to a task with a JSON schema is not necessarily required, certainly not for a first implementation. I think it's a topic to be discussed if we'd want it at all.

**Comment by jluethi** [comment]
Revisiting old issues & closing this now

Current usage experience has shown that we don't need to add parameters to existing tasks. We can either edit the schema when needed or add a new version of the task
